c++/include/orc/OrcFile.hh

```diff
@@ -58,6 +73,17 @@ namespace orc {
      */
     virtual void read(void* buf, uint64_t length, uint64_t offset) = 0;
 
+    /**
+     * Read data asynchronously.
+     * @param offset the position in the stream to read from.
+     * @param length the number of bytes to read.
+     * @return a future that will be set to the buffer when the read is complete.
+     */
+    virtual std::future<BufferPtr> readAsync(uint64_t /*offset*/, uint64_t /*length*/,
```

Comment:
readAsync is an asynchronous process, which means `DataBuffer` is constructed and filled asynchronously in the future. If we construct `DataBuffer` in the main thread in which `readAsync` is invoked, the main thread may be blocked by page faults. It is not friendly to performance.

Reply:
I don't think this is a good practice. How to allocate the buffer is an implementation detail. In the current implementation, the DataBuffer instance can be managed by ReadRangeCache. So I would prefer the following signature:
```
virtual std::future<void> read(void* buf, uint64_t length, uint64_t offset) = 0;
```

Reply:
How to allocate buffer is only related to MemoryPool implementation. If we use the new interface, the cost of allocating will occupy the main thread instead of backgroud thread. It is not good to performance.

Reply:
There is an extra advantage if we use the old interface: we could easily know `readAsync` is success or not by `BufferPtr != nullptr`

Reply:
Do you have the number of the performance penalty? 

We need to trade off between performance and usability. In apache/arrow, its input stream supports both styles of returning a buffer or copying into a provided buffer. If we need to return a buffer in the readAsync function, perhaps we need to add an equivalent sync function to return a buffer as well.

> There is an extra advantage if we use the old interface: we could easily know readAsync is success or not by BufferPtr != nullptr

Isn't an exception expected if the read is unsuccessful?

Reply:
I added a benchmark about it. 
``` cpp

static constexpr size_t numThreads = 16;
static constexpr size_t bufferSize = 1024 * 1024;
static std::vector<std::thread> threads1(numThreads);
static std::vector<std::thread> threads2(numThreads);

static void allocateAndWriteInBackground()
{
    for (size_t i = 0; i < numThreads; ++i)
    {
        threads1[i] = std::thread(
            []()
            {
                orc::DataBuffer<char> buffer(*orc::getDefaultPool(), bufferSize);
                for (size_t j = 0; j < bufferSize; ++j)
                    buffer[j] = static_cast<char>(j);
            });
    }

    for (auto & thread : threads1)
    {
        thread.join();
    }
}

static void BM_allocateAndWriteInBackground(benchmark::State & state)
{
    for (auto _ : state)
        allocateAndWriteInBackground();
}

static void writeInBackground()
{
    std::vector<std::shared_ptr<orc::DataBuffer<char>>> buffers(numThreads);
    for (size_t i = 0; i < numThreads; ++i)
    {
        auto buffer = std::make_shared<orc::DataBuffer<char>>(*orc::getDefaultPool(), bufferSize);
        buffers[i] = buffer;
        threads2[i] = std::thread(
            [&buffer]()
            {
                for (size_t j = 0; j < bufferSize; ++j)
                    (*buffer)[j] = static_cast<char>(j);
            });
    }

    for (auto & thread : threads2)
    {
        thread.join();
    }
}

static void BM_writeInBackground(benchmark::State & state)
{
    for (auto _ : state)
        writeInBackground();
}

BENCHMARK(BM_allocateAndWriteInBackground);
BENCHMARK(BM_writeInBackground);
BENCHMARK_MAIN();
```

``` bash
Running ./build_gcc/src/Columns/benchmarks/column_insert_many_from
Run on (32 X 2100 MHz CPU s)
CPU Caches:
  L1 Data 32 KiB (x16)
  L1 Instruction 32 KiB (x16)
  L2 Unified 1024 KiB (x16)
  L3 Unified 11264 KiB (x2)
Load Average: 11.87, 43.43, 55.11
--------------------------------------------------------------------------
Benchmark                                Time             CPU   Iterations
--------------------------------------------------------------------------
BM_allocateAndWriteInBackground    2122695 ns       631300 ns         1121
BM_writeInBackground               3096734 ns       529098 ns         1367
```

Reply:
Remind that user could use the buffer when it's
`virtual std::future<void> read(void* buf, uint64_t length, uint64_t offset) = 0;`

Reply:
In this pr, the buffers are all members of ReadRangeCache. The interface change doesn't influence the behavior of buffer reusing.

Reply:
Do you think providing both interfances is ok ?

Reply:
@wgtmac @ffacs I had renamed the old interface to `readAsyncInternal` and moved it under `protected`. User could not invoke it. What user could see is the new interface `readAsync`. And it has a default implementation. Users could override it if they are not satisfied with it.

Reply:
TBH, I don't think this penalty is unacceptable. In practice, the number of issued I/Os should not be too many after being coalesced. When the data is in a hot cache (e.g. SSD or memory), this feature may not improve too much and may get worse when predicate pushdown is effective (due to reading more chunks than needed). It is super helpful when data is in the cloud object store, in which case this penalty is trivial.